packages/protocol/contracts/stability/GrandaMento.sol

```diff
@@ -180,6 +218,134 @@ contract GrandaMento is
+   * @dev Sender must be the approver. Exchange proposal must be in the Proposed state.
+   * @param proposalId The identifier of the proposal to approve.
+   */
+  function approveExchangeProposal(uint256 proposalId) external onlyApprover {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Ensure the proposal is in the Proposed state.
+    require(proposal.state == ExchangeProposalState.Proposed, "Proposal must be in Proposed state");
+    // Set the time the approval occurred and change the state.
+    proposal.approvalTimestamp = block.timestamp;
+    proposal.state = ExchangeProposalState.Approved;
+    emit ExchangeProposalApproved(proposalId);
+  }
+
+  /**
+   * @notice Cancels an exchange proposal.
+   * @dev Only callable by the exchanger if the proposal is in the Proposed state
+   * or the owner if the proposal is in the Approved state.
+   * @param proposalId The identifier of the proposal to cancel.
+   */
+  function cancelExchangeProposal(uint256 proposalId) external nonReentrant {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Require the appropriate state and sender.
+    // This will also revert if a proposalId is given that does not correspond
+    // to a previously created exchange proposal.
+    require(
+      (proposal.state == ExchangeProposalState.Proposed && proposal.exchanger == msg.sender) ||
+        (proposal.state == ExchangeProposalState.Approved && isOwner()),
+      "Sender cannot cancel the exchange proposal"
+    );
+    // Mark the proposal as cancelled.
+    proposal.state = ExchangeProposalState.Cancelled;
+    // Get the token and amount that will be refunded to the proposer.
+    (IERC20 refundToken, uint256 refundAmount) = getSellTokenAndSellAmount(proposal);
+    // Finally, transfer out the deposited funds.
+    require(
+      refundToken.transfer(proposal.exchanger, refundAmount),
+      "Transfer out of refund token failed"
+    );
+    emit ExchangeProposalCancelled(proposalId, msg.sender);
+  }
+
+  /**
+   * @notice Executes an exchange proposal that's been approved and not vetoed.
+   * @dev Callable by anyone. Reverts if the proposal is not in the Approved state
+   * or vetoPeriodSeconds has not elapsed since approval.
+   * @param proposalId The identifier of the proposal to execute.
+   */
+  function executeExchangeProposal(uint256 proposalId) external nonReentrant {
+    ExchangeProposal storage proposal = exchangeProposals[proposalId];
+    // Require that the proposal is in the Approved state.
+    require(proposal.state == ExchangeProposalState.Approved, "Proposal must be in Approved state");
+    // Require that the veto period has elapsed since the approval time.
+    require(
+      proposal.approvalTimestamp.add(vetoPeriodSeconds) <= block.timestamp,
+      "Veto period not elapsed"
+    );
+    // Mark the proposal as executed.
+    proposal.state = ExchangeProposalState.Executed;
```

Comment:
nit: to make debugging more straightforward I'dd place this right before the `emit`

Reply:
I changed this but thought about it some more and I think updating the state before doing any token transfers/minting/burning is the right move. Doing so is a common practice against any kind of reentrancy, which is ofc prevented by the `nonReentrant` modifier, but just to err on the side of safety may be good to keep